File: src/main/scala/coda/CoreTypes.scala
package coda

import com.oracle.truffle.api.dsl.{ ImplicitCast, TypeSystem }
import java.math.BigInteger

@TypeSystem(
  Array(
    classOf[Boolean],
    classOf[Float],
    classOf[Double],
    classOf[Int],
    classOf[Long],
    classOf[Short],
    classOf[Byte],
    classOf[Char],
    classOf[BigInteger],
    classOf[String]
  )
)
class CoreTypes {}
object CoreTypes {
  // used so we can have an unboxed long-based fast path for "small" big integer computations
  @ImplicitCast
  def asBigInteger(value: Long): BigInteger = BigInteger.valueOf(value)
}

File: src/main/java/core/nodes/CoreTypes.java
package core.nodes;

import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
import com.oracle.truffle.api.dsl.ImplicitCast;
import com.oracle.truffle.api.dsl.TypeSystem;
import core.values.CoreBigInteger;
import java.math.BigInteger;

@TypeSystem({
  boolean.class,
  int.class,
  long.class
})
public abstract class CoreTypes {
  @ImplicitCast
  @TruffleBoundary
  public static CoreBigInteger castBigInteger(long value) {
    return new CoreBigInteger(BigInteger.valueOf(value));
  }
}
